feat(api): JSON protocol labware loading support in Protocol Engine

amitlissack opened the pull request:
# Overview

JSON Protocols contain full labware definitions and labware ids in their body. This PR provides a JSON protocol runner the ability to send commands to the ProtocolEngine to add labware definitions and load labware with a specific id. 

closes #7429 

# Changelog

- add AddLabwareDefinitionRequest: a command that submits a labware definition to the protocol engine.
- add AddLabwareDefinitionResult: the response to AddLabwareDefinitionRequest containing the `loadName`, `namespace`, and `version` of the uploaded definition.
- The labware state store handles `AddLabwareDefinitionResult` to store the labware definition in a new mapping of URI to LabwareDefintion.
- Added `get_labware_definition` to Labware State store.
- `LabwareDefinitions` will not be loaded by equipment handler if an instance of that definition already exists in the state store. 
- Use pydantic `LabwareDefinition` instead of the typedict.
- Tavern tests to exercise `LoadLabwareRequest` with `labwareId` provided.


# Review requests



# Risk assessment

None

Commit:
AddLabwareDefinitionRequest implementation

Commit:
lint errors.

Commit:
PE uses LabwareDefinition model.

Commit:
lint

Commit:
clean up

Commit:
fixed labware definition is included in cache.

Commit:
rename some models.

Commit:
LoadLabwareRequest accepts optional labwareId to bypass labware id generation.

amitlissack commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -52,15 +52,37 @@ async def load_labware(
         namespace: str,
         version: int,
         location: LabwareLocation,
+        labware_id: Optional[str]
```
Should this default to `None`?

mcous commented:
Was thinking through this, and I think I see a potential snag with this model: if a JSON protocol must be collapsed into a (1) load definition command and then (2) a load labware command that depends on the output of the load definition execution (because it breaks apart the URI). Because the input of (2) depends on the execution output of (1), we can no longer load all protocol commands without executing anything (in a pedantic sense).

Not sure what the resolution is here, but a few ideas in order of least annoying to most annoying

1. Accept that it's ok for the command translator to know how to break apart a labware URI into load labware request params, move on with our lives
2. Create some sort of "load labware by URI" command
3. Fallback to load labware by definition command

Sitting with it for a few hours, I'm pretty ok with (1), which I think means this PR is all good, but figured I should check in

mcous commented on api/src/opentrons/protocol_engine/commands/add_labware_definition.py:
```diff
@@ -0,0 +1,54 @@
+
+class AddLabwareDefinitionRequest(BaseModel):
+    """Request to add a labware definition."""
+    labware_definition: LabwareDefinition = Field(
```
Should this be camel case? Also might be nice to just shorten to `definition` to match the `LoadLabwareResult`

mcous commented on api/src/opentrons/protocol_engine/commands/load_labware.py:
```diff
@@ -26,6 +29,11 @@ class LoadLabwareRequest(BaseModel):
         description="The labware definition version.",
     )
+    labwareId: Optional[str] = Field(
+        ...,
```
Should this default to `None`?

amitlissack replied:
Yes

mcous commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -52,15 +52,37 @@ async def load_labware(
-        )
+        try:
+            # Try to use existing definition in state.
+            definition = self._state.labware.get_labware_definition(
```
Shoot, this name sort of conflicts with `state.labware.get_definition(labware_id=labware_id)` added in #7769. Should we shift the names of one or both of these to be a little more clear?

amitlissack replied:
Yes. I'll fix after we merge yours.

mcous commented on api/src/opentrons/protocol_engine/state/labware.py:
```diff
@@ -124,8 +162,24 @@ def __init__(self, deck_fixed_labware: Sequence[DeckFixedLabware]) -> None:
+                load_name=command.request.loadName,
+                version=command.request.version
+            )
+            self._state._labware_definitions_by_uri[uri] = \
```
This is storing the same definition in multiple places in the same state tree. We should strive to avoid (ideally) any sort of state duplication, because it makes debugging and source of truth much harder to follow.

Why are we storing the definition upon completion of `LoadLabwareResult`?

amitlissack replied:
Good point on the first.

The second, just to cache the definition so we don't need to reload definitions.

mcous reviewed (changes requested):
I think we need to avoid putting the definitions in multiple places in the labware store, otherwise we're going to run into debugging / source of truth issues.

Also, I just pushed up a PR that adds a pretty similar method to the labware store, so let's figure out how to name both of them to avoid confusion

amitlissack commented:
> Was thinking through this, and I think I see a potential snag with this model: if a JSON protocol must be collapsed into a (1) load definition command and then (2) a load labware command that depends on the output of the load definition execution (because it breaks apart the URI). Because the input of (2) depends on the execution output of (1), we can no longer load all protocol commands without executing anything (in a pedantic sense).
> 
> Not sure what the resolution is here, but a few ideas in order of least annoying to most annoying
> 
> 1. Accept that it's ok for the command translator to know how to break apart a labware URI into load labware request params, move on with our lives
> 2. Create some sort of "load labware by URI" command
> 3. Fallback to load labware by definition command
> 
> Sitting with it for a few hours, I'm pretty ok with (1), which I think means this PR is all good, but figured I should check in

Since we call `LoadLabwareRequest` knowing the components of the URI, it's not really necessary to use the values in `AddLabwareDefinitionResult`. 

So I am ok with `1` too.

Commit:
rebase fixes.

Commit:
rename get_definition to get_labware_definition.
rename AddLabwareDefinitionRequest.labware_definition to definition.
AddLabwareDefinitionRequest.labwareId defaults to None.
Change get_labware_definition to get_definition_by_uri.
LabwareData holds uri instead of definition.
redo test_geometry_state using Decoy.

mcous commented on api/src/opentrons/protocol_api_experimental/labware.py:
```diff
@@ -92,10 +92,10 @@ def load_name(self) -> str:
             labware_id=self._labware_id
         )
-        return definition["parameters"]
+        return cast(LabwareParameters, definition.parameters.dict())
```
Oh boy, this whole thing with the type dicts. I wonder how much of a "breaking" change it is to switch to the models. I also wonder if this `parameters` method is really even used

amitlissack replied:
This will be version 3. We can break away. Can't we?

mcous commented on api/src/opentrons/protocol_engine/state/labware.py:
```diff
@@ -133,8 +161,24 @@ def __init__(self, deck_fixed_labware: Sequence[DeckFixedLabware]) -> None:
+                uri=uri,
                 calibration=command.result.calibration,
             )
+        elif isinstance(command.request, AddLabwareDefinitionRequest) and \
```
I believe this check on `command.request` is redundant. The system will only ever pair a `AddLabwareDefinitionRequest` with a  `AddLabwareDefinitionResult`, and the typing should know that

amitlissack replied:
It is. It's just there for mypy.

amitlissack replied:
I'll double check but I added the check for mypy.

amitlissack replied:
Ha! You were right. Weirdly pycharm marked it as a type error but not mypy. I don't get it. But I removed the check

mcous commented on api/tests/opentrons/conftest.py:
```diff
@@ -506,7 +509,9 @@ def minimal_labware_def():
             "loadName": "minimal_labware_def",
-            "quirks": ["a quirk"]
+            "isMagneticModuleCompatible": True,
+            "quirks": ["a quirk"],
```
I wonder if we can remove `quirks` without breaking any tests. It's not required, according to the JSON schema

amitlissack replied:
I am guessing it's in use by some test. I'll check.

mcous commented on robot-server/tests/conftest.py:
```diff
@@ -234,6 +234,63 @@ def _get_labware_fixture(fixture_name):
 
 
+@pytest.fixture
+def minimal_labware_def():
```
Maybe in the future we can create a fixtures library in `shared-data` (or wherever) and import them into `conftest.py`'s instead

amitlissack replied:
That's a nice idea. It's not really clear to me what does and does not belong in shared-data.

Commit:
remove redundant type check.
